### PR TITLE
Add script for hyperparameter search

### PR DESCRIPTION
**Pull request by Eleven1Liu**
**Description**: 
(1) Add hyperparameter script and config with all parameters (`caml_tune.yml`)
- Search algorithms: `OptunaSearch`, `BayesOptSearch`, and `BasicVariantGenerator` (random choice or grid search)
- Optional requirements
  - ray >= 1.3.0
  - optuna == 2.6.0
  - bayesian-optimization
- configuration (`caml_tune.yml`): we need a configuration with all args now to run without `main.py`.

(2) Replace `init_env` with  `set_seed` and `init_device` in `utils.py`.

**Usage**
(1) set values in the config files
```yaml
learning_rate: ['choice', [0.001, 0.003, 0.0001, 0.0003]]
num_filter_per_size:  ['choice', [50, 150, 250, 350, 450, 550]]
filter_sizes: [ ['choice', [2, 4, 6, 8, 10]]]
dropout:  ['choice',  [0.2, 0.4, 0.6, 0.8]]
```

**or**
```yaml
learning_rate: ['uniform', [0.0001, 0.001]]
dropout:  ['uniform',  [0.2, 0.8]]
```

(2) run with:
```python
python search_params.py  --config example_config/MIMIC-50/caml_tune.yml 
                         --search_alg random
                         --search_params dropout learning_rate filter_sizes num_filter_per_size
```

### CODE REVIEW

File: libmultilabel/utils.py [hunk 1]
```diff
@@ -63,6 +64,13 @@ def time(self):
 
 
 def dump_log(config, metrics, split):
+    """Write log includes config and the evaluation scores.
```
Comment (cjlin1) [inline libmultilabel/utils.py:67]:
```suggestion
    """Write log including config and the evaluation scores.
```

File: main.py [hunk 2]
```diff
@@ -93,27 +91,10 @@ def get_config():
...
-        config.device = torch.device('cuda')
-    else:
-        config.device = torch.device('cpu')
-        # https://github.com/pytorch/pytorch/issues/11201
-        torch.multiprocessing.set_sharing_strategy('file_system')
-    logging.info(f'Using device: {config.device}')
+def main():
+    config = get_config()
+    set_seed(seed=config.seed)
+    config.device = init_device(config.cpu)
```
Comment (cjlin1) [inline main.py:97]:
```suggestion
    config.device = init_device(use_cpu=config.cpu)
```
config.cpu sounds like we want to use cpu.. But "use_cpu" shows this means to use it or not

File: search_params.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        assert len(values) == 2 and values[0] < values[1]
+        # sample an option uniformly between values[0] and values[1]
+        return tune.uniform(values[0], values[1])
+    else:
+        # sample an option uniformly from list of values
+        return tune.choice(values)
+
+
+def get_search_algorithm(search_alg, metric=None, mode=None):
+    """Specify a search algorithm. You should pip install this search algorithm first.
```
Comment (cjlin1) [inline search_params.py:73]:
```suggestion
    """Specify a search algorithm and you must pip install it first.
```

File: search_params.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    See more details here: https://docs.ray.io/en/master/tune/api_docs/suggestion.html
```
Comment (cjlin1) [inline search_params.py:74]: So users need to check this address for available algs? If so, should mention it in README

File: search_params.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
+import argparse
+import logging
+import os
+import time
+import yaml
+from datetime import datetime
+from pathlib import Path
+
+from ray import tune
```
Comment (cjlin1) [inline search_params.py:9]: add ray in requirements.txt?

File: example_config/MIMIC-50/caml_tune.yml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+data_name: MIMIC-50
+min_vocab_freq: 3
+max_seq_length: 2500
+
+# train
+seed: 1337
+epochs: 200
+batch_size: 16
+optimizer: adam
+learning_rate: ['choice', [0.001, 0.003, 0.0001, 0.0003]]
```
Comment (cjlin1) [inline example_config/MIMIC-50/caml_tune.yml:12]: So "choice" means this parameter will be tuned? Is this a keyword designed by us or required by ray? Also this should be mentioned in README

File: example_config/MIMIC-50/caml_tune.yml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+metrics_thresholds: [0.5]
+momentum: 0.9
+predict_out_path: null
+result_dir: runs/optuna-50
+test_path: data/MIMIC-50/test.txt
+train_path: data/MIMIC-50/train.txt
+val_path: null
+val_size: 0.2
+vocab_file: data/MIMIC-50/vocab.csv
+vocab_label_map: null
```
Comment (cjlin1) [inline example_config/MIMIC-50/caml_tune.yml:52]: Why do we need to specify values of these parameters here?? We didn't have to do that in caml.yaml

File: libmultilabel/utils.py [hunk 8]
```diff
@@ -98,3 +108,30 @@ def save_top_k_predictions(class_names, y_pred, predict_out_path, k=100):
...
+            torch.set_deterministic(True)
+            torch.backends.cudnn.benchmark = False
+        else:
+            logging.warning(
+                f'the random seed should be a non-negative integer')
+
+
+def init_device(use_cpu=False):
+    if not use_cpu and torch.cuda.is_available():
+        # set a debug environment variable CUBLAS_WORKSPACE_CONFIG to ":16:8" (may limit overall performance) or ":4096:8" (will increase library footprint in GPU memory by approximately 24MiB).
```
Comment (cjlin1) [inline libmultilabel/utils.py:128]:
```suggestion
        # Set a debug environment variable CUBLAS_WORKSPACE_CONFIG to ":16:8" (may limit overall performance) or ":4096:8" (will increase library footprint in GPU memory by approximately 24MiB).
```
it seems the python guide suggests that comments are "full sentence and the 1st word capitalized" 
We can gradually change other places later

File: search_params.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    classes = data_utils.load_or_build_label(model_config, datasets)
+
+    model = Model(model_config, word_dict, classes)
+    model.train(datasets['train'], datasets['val'])
+    model.load_best()
+
+    # return best eval metric
+    dev_loader = data_utils.get_dataset_loader(model_config, datasets['val'], model.word_dict, model.classes, train=False)
+    results = evaluate(model_config, model, dev_loader, split='val', dump=True)
+    yield results
```
Comment (cjlin1) [inline search_params.py:33]: we use yield instead of return because this fun is called/resumed several times?

File: search_params.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+    datasets = data_utils.load_datasets(model_config)
+    word_dict = data_utils.load_or_build_text_dict(model_config, datasets['train'])
+    classes = data_utils.load_or_build_label(model_config, datasets)
+
+    model = Model(model_config, word_dict, classes)
+    model.train(datasets['train'], datasets['val'])
+    model.load_best()
+
+    # return best eval metric
+    dev_loader = data_utils.get_dataset_loader(model_config, datasets['val'], model.word_dict, model.classes, train=False)
```
Comment (cjlin1) [inline search_params.py:31]: should val_loader following our convention now

File: search_params.py (hunk 5 above)
Comment (Eleven1Liu) [reply search_params.py:9]: Yes, done.

File: README.md [hunk 11]
```diff
@@ -144,3 +144,20 @@ python3 main.py --eval --config CONFIG_PATH --load_checkpoint CHECKPOINT_PATH --
 - Use `--save_k_predictions` to save the top K predictions for each instance in the test set. K=100 if not specified.
 - Use `--predict_out_path` to specify the file for storing the predicted top-K labels/scores.
 
+
+### Hyperparameter Search
+We leverage Ray Tune, which is a python library for hyperparameter tuning, to select parameters. We provide a program `search_params.py` to demonstrate how to run LibMultiLabel with Ray Tune. An example to run it as follows.
+```
+python search_params.py  --config example_config/MIMIC-50/caml_tune.yml
+                         --search_alg random
```
Comment (cjlin1) [inline README.md:152]: Can this --search_alg be specified in the yml file?

**Review by cjlin1 — APPROVED** [review]

**Pull request merged**